Build Standard entirely against Arcade, without Buildtools

wtgodbe opened the pull request:
This cuts out buildtools entirely from the Standard repo, so that we fully rely on the Arcade SDK to build against instead. It replaces the functionality for code analyzers, `ValidateApiCompatForSrc`, and the build tasks.

@ericstj @safern @danmosemsft @ViktorHofer @chcosta PTAL

CC @karelz

wtgodbe commented:
TODO: 

- [ ] Check why `netstandard.dll` in `artifacts\bin\netstandard\Debug\net461` is much smaller than the one that used to be in `bin\ref\netstandard\2.0.0.0\net461`
- [ ] Get `ValidateApiCompatForSrc` running

wtgodbe commented:
With the latest changes, I'm getting errors like the following:

> E:\code\standard\artifacts\obj\netstandard\Debug\xamarin.tvos\netstandard.forwards.cs(922,87): error CS0234: The type or namespace name 'ISOWeek' does not exist in the namespace 'System.Globalization' (are you missing an assembly reference?) [E:\code\standard\netstandard\src\netstandard.csproj]

@terrajobst I see that you recently added `ISOWeek` to `System.Globalization`, and I've confirmed that I have that change locally: https://github.com/dotnet/standard/pull/963/files. Do you know what I might be missing to get CSC to consolidate that reference?

wtgodbe commented:
@terrajobst I confirmed last night that I had unintentionally turned off generation of `netstandard.forwards.cs` a couple PRs back, so that's why we didn't catch this before. I've re-enabled generating that file with this PR. Will update the `GenApi.exclude.*.txt` files. Thanks!

wtgodbe commented:
Due to some changes I made last month, we (unintentionally) haven't been running APICompat on the xamarin/net461 baselines for a while. This PR includes updates to the baselines with the `CannotRemoveAttribute`, `CannotChangeAttribute`, and `TypesMustExist` checks. CC @ericstj @terrajobst

wtgodbe commented:
This is now ready for review @safern @ericstj @ViktorHofer @chcosta

danmosemsft commented on netstandard/src/ApiCompatBaseline.net461.txt:
```diff
@@ -5,6 +5,8 @@ MembersMustExist : Member 'System.Array.Fill<T>(T[], T)' does not exist in the i
 MembersMustExist : Member 'System.Array.Fill<T>(T[], T, System.Int32, System.Int32)' does not exist in the implementation but it does exist in the contract.
 MembersMustExist : Member 'System.Array.Reverse<T>(T[])' does not exist in the implementation but it does exist in the contract.
 MembersMustExist : Member 'System.Array.Reverse<T>(T[], System.Int32, System.Int32)' does not exist in the implementation but it does exist in the contract.
+CannotRemoveAttribute : Attribute 'System.Reflection.DefaultMemberAttribute' exists on 'System.ArraySegment<T>' in the contract but not the implementation.
```
Why are you getting all these new entries?

wtgodbe replied:
They're new checks as of https://github.com/dotnet/arcade/commit/c6e4b4bb3e3aab02e542d16891295fe18c466def, and we haven't been running APICompat on these guys since the checks were added

wtgodbe commented:
Hmm, the build failures aren't reproing for me locally:

> artifacts\obj\netstandard\Release\xamarin.ios\netstandard.forwards.cs(126,73): error CS0234: The type or namespace name 'IAsyncDisposable' does not exist in the namespace 'System' (are you missing an assembly reference?) [F:\vsagent\34\s\netstandard\src\netstandard.csproj]

Will try to see if I can get a repro & investigate

danmosemsft commented on eng/depProj.targets:
```diff
@@ -0,0 +1,146 @@
+<?xml version="1.0" encoding="utf-8"?>
```
This looks the same as the file in CoreFX. Should we have some central place we share this, presumably Arcade?

In general is the eng folder here and in CoreFX a mixture of Arcade stuff and repo stuff, there's not separate folders?

ViktorHofer replied:
Right, we knew that there will be a few duplicates as we don't want to move everything into arcade. Maybe we can either mirror these files manually (does the mirror support single files?), find another place where they can live or just live with the duplicates for now.

danmosemsft commented on eng/Tools.props:
```diff
@@ -2,51 +2,90 @@
 
-    <Touch Files="$(BuildToolsSemaphore)" AlwaysCreate="true" />
-    <OnError ExecuteTargets="DumpInitToolsLog" />
+    <Copy 
```
SkipUnchangedFiles=true parameter on <Copy>

danmosemsft commented on build.proj:
```diff
@@ -16,15 +16,7 @@
-
-    <MSBuild Projects="@(_RestoreProjects)" />
-  </Target>
+   <Target Name="Restore"/>
```
indenting off

danmosemsft commented on Directory.Build.props:
```diff
@@ -42,21 +41,26 @@
-    <ExcludeVersioningImport>true</ExcludeVersioningImport>
-    <ExcludePackagingImport>true</ExcludePackagingImport>
-    <ExcludePackageLibsImport>true</ExcludePackageLibsImport>
+    <LanguageTargets Condition="'$(LanguageTargets)' == '' AND '$(MSBuildProjectExtension)' != '.csproj' AND '$(MSBuildProjectExtension)' != '.vbproj' AND '$(MSBuildProjectExtension)' != '.fsproj'">$(MSBuildToolsPath)\Microsoft.Common.targets</LanguageTargets>
```
Nit, the convention seems to be lower case 'and' and 'or'

also below

ViktorHofer commented on Directory.Build.targets:
```diff
@@ -17,30 +17,43 @@
+  Hack workaround for .depproj's built with TargetFramework=net46
+  This causes us to unnecessarily attempt to restore framework assemblies from the project directory
+  This target clears out those assemblies from the list of assets to resolve -->
+  <Target Name="_ClearResolvedFrameworkAssemblies"
```
I currently see 3 depproj files in the standard repo. Which one has a configuration for net46?

wtgodbe replied:
Ah, the comment should say "net47": https://github.com/dotnet/standard/blob/master/apicompat/netstandard1x/netstandard.builds#L29-L31

ViktorHofer commented on Directory.Build.targets:
```diff
@@ -17,30 +17,43 @@
 
+  <Import Project="$(RepositoryEngineeringDir)CodeAnalysis.targets" />
+
+  <Import Project="$(MSBuildThisFileDirectory)eng/depProj.targets" Condition="'$(MSBuildProjectExtension)' == '.depproj'" />
```
```suggestion
  <Import Project="$(RepositoryEngineeringDir)depProj.targets" Condition="'$(MSBuildProjectExtension)' == '.depproj'" />
```

ViktorHofer commented on apicompat/netstandard1x/netstandard1x.depproj:
```diff
@@ -1,14 +1,12 @@
   <PropertyGroup>
     <NuGetDeploySourceItem>Reference</NuGetDeploySourceItem>
-    <NuGetTargetMoniker Condition="'$(NuGetTargetMoniker)'==''">.NETStandard,Version=v1.6</NuGetTargetMoniker>
     <TargetFramework Condition="'$(TargetFramework)'==''">netstandard1.6</TargetFramework>
```
Do we really need a default value here and if yes should we update it to netstandard2.0 or 2.1?

wtgodbe replied:
Right now we just build this for netstandard1.0-1.6 & net47. I'll chat with @terrajobst later today about if we should baseline for 2.0/2.1 as well.

wtgodbe replied:
https://github.com/dotnet/standard/issues/1029

ViktorHofer commented on apicompat/netstandard1x/netstandard1x.depproj:
```diff
@@ -1,14 +1,12 @@
+<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <NuGetDeploySourceItem>Reference</NuGetDeploySourceItem>
-    <NuGetTargetMoniker Condition="'$(NuGetTargetMoniker)'==''">.NETStandard,Version=v1.6</NuGetTargetMoniker>
```
As this project is now sdk based I don't think we need the NuGetTargetMonikers anymore in https://github.com/dotnet/standard/blob/master/apicompat/netstandard1x/netstandard.builds#L30.

wtgodbe replied:
Good call, I'll remove that

wtgodbe commented:
@ViktorHofer I think that's a good idea, but I'd rather address it in a separate PR. I can put up an issue for myself to cut out .depproj's once this is merged

ViktorHofer commented:
I agree. It would be great if we can make the standard repo behave like a normal repo :)

wtgodbe commented:
The arguments to CSC appear to be the same on my machine & in the CI. Still looking into what might be causing the errors.

wtgodbe commented:
The types/namespaces referenced in the errors are:

> IAsyncDisposable, DynamicILInfo, EventToken, FieldToken, MethodToken, ParameterToken, PropertyToken, SignatureToken, StringToken, TypeToken

None of these appear in my local `netstandard.forwards.cs` files, which leads me to believe there's some exclusion list that isn't being honored? Still looking.

wtgodbe commented:
Ah, looks like there were some recent changes that I don't have locally, such as https://github.com/dotnet/standard/pull/1014. @terrajobst do I need to add these new additions to the GenApi exclusion lists?

wtgodbe commented:
I've addressed feedback (https://github.com/dotnet/standard/pull/1027/commits/2b621314965a7d86bf6496f3419e106004c7be2c) & fixed the build breaks, @ViktorHofer does everything look good now?

wtgodbe commented:
@dotnet-bot test this please

ViktorHofer commented on Directory.Build.props:
```diff
@@ -186,6 +190,8 @@
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
   </PropertyGroup>
 
+  <Import Project="$(MSBuildThisFileDirectory)eng\ReferenceAssemblies.props" />
```
```suggestion
  <Import Project="$(RepositoryEngineeringDir)ReferenceAssemblies.props" />
```

ViktorHofer commented on Directory.Build.props:
```diff
@@ -12,7 +12,6 @@
   </PropertyGroup>
 
   <Import Condition="Exists('..\Directory.Build.props')" Project="..\Directory.Build.props" />
```
what is this for?

wtgodbe replied:
I think it's leftover from the old dir.props model, it used to be convention to always upload a higher-level dir.props if it existed. I'll delete it.

ViktorHofer commented on Directory.Build.props:
```diff
@@ -27,8 +26,8 @@
 
```
In line 22 `RootIntermediateOutputPath` is still defined. I thought with the buildtools removal we don't need it anymore?

wtgodbe replied:
Good catch, I'll delete it

ViktorHofer commented on Directory.Build.targets:
```diff
@@ -17,30 +17,43 @@
 
-  <Import Project="$(ToolsDir)/Build.Common.targets" Condition="Exists('$(ToolsDir)/Build.Common.targets')" />
-
   <Import Project="eng\versioning.targets" />
```
Only if you are touching the PR again:

```suggestion
  <Import Project="$(RepositoryEngineeringDir)versioning.targets" />
```

ViktorHofer commented on Directory.Build.targets:
```diff
@@ -17,30 +17,43 @@
+
   <Import Project="$(ToolSetCommonDirectory)Tools.proj.nuget.g.targets" Condition="Exists('$(ToolSetCommonDirectory)Tools.proj.nuget.g.targets')" />
 
   <!-- Define this now until we can clean-up targets that depend on it in the packaging targets -->
```
does that still apply with the packaging from arcade?

ViktorHofer commented on apicompat/net461/net461.depproj:
```diff
@@ -1,12 +1,11 @@
-    <NuGetTargetMoniker>.NETFramework,Version=v4.6.1</NuGetTargetMoniker>
-    <NuGetTargetMonikerShort>net461</NuGetTargetMonikerShort>
+    <TargetFramework>net461</TargetFramework>
     <OutputPath>$(RestoredRefRootPath)net461</OutputPath>
```
Only if you are touching this PR again:

```suggestion
    <OutputPath>$(RestoredRefRootPath)$(TargetFramework)</OutputPath>
```

ViktorHofer commented on build.proj:
```diff
@@ -16,15 +16,7 @@
-
-    <MSBuild Projects="@(_RestoreProjects)" />
-  </Target>
+  <Target Name="Restore"/>
```
You now also need this change: https://github.com/dotnet/corefx/pull/34168/files#diff-991506ab3bfb2ea4767ed342c4f95726R5

ViktorHofer replied:
It disables calling directly into the NuGet.targets restore target if the repository build project does not use NuGet restore.

ViktorHofer replied:
https://github.com/dotnet/corefx/issues/34164

ViktorHofer commented:
Ok found another left-over which probably should have been part of this PR: https://github.com/dotnet/standard/blob/master/eng/configure-toolset.ps1#L1. It seems now without buildtools we don't need this file and its Unix counterpart anymore.
